source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,274 @@
+    return config_->contentTypeValues().find(value) != config_->contentTypeValues().end();
+  }
+  return true;
+}
+
+// TODO(gsagula): It seems that every proxy has a different opinion how to handle Etag. Some
+// discussions around this topic have been going on for over a decade, e.g.,
+// https://bz.apache.org/bugzilla/show_bug.cgi?id=45023
+// This design attempts to stay more on the safe side by preserving weak etags and removing
+// the strong ones when disable_on_etag_header is false. Envoy does NOT rewrite Etags.
+bool GzipFilter::isEtagAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* etag = headers.Etag();
+  if (config_->disableOnEtagHeader()) {
+    return !etag;
+  }
+
+  if (etag) {
+    absl::string_view value(etag->value().c_str());
+    if (value.length() > 2 && !((value[0] == 'w' || value[0] == 'W') && value[1] == '/')) {
+      headers.removeEtag();
+    }
+  }
+
+  return true;
+}
+
+bool GzipFilter::isLastModifiedAllowed(HeaderMap& headers) const {
+  return !(config_->disableOnLastModifiedHeader() && headers.LastModified());
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_length = headers.ContentLength();
+  if (content_length) {
+    uint64_t length;
+    return StringUtil::atoul(content_length->value().c_str(), length) &&
+           length >= config_->minimumLength();
+  }
+
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  return (transfer_encoding &&
+          StringUtil::findToken(transfer_encoding->value().c_str(), ",",
+                                Http::Headers::get().TransferEncodingValues.Chunked.c_str()));
+}
+
+bool GzipFilter::isTransferEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  if (transfer_encoding) {
+    for (auto header_value :
+         StringUtil::splitToken(transfer_encoding->value().c_str(), ",", true)) {
+      const auto trimmed_value = StringUtil::trim(header_value);
+      if (trimmed_value == Http::Headers::get().TransferEncodingValues.Gzip ||
+          trimmed_value == Http::Headers::get().TransferEncodingValues.Deflate) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+void GzipFilter::insertVaryHeader(HeaderMap& headers) {
+  const Http::HeaderEntry* vary = headers.Vary();
+
+  if (vary && !StringUtil::findToken(vary->value().c_str(), ",",
+                                     Http::Headers::get().VaryValues.AcceptEncoding, true)) {
+    std::string new_header;
+
+    // separator's length: ", "
+    const uint64_t comma_separator_len{2};
+    new_header.reserve(strlen(vary->value().c_str()) +
+                       Http::Headers::get().VaryValues.AcceptEncoding.size() + comma_separator_len);
+
+    new_header.append(vary->value().c_str());
+    new_header.append(", ");
+    new_header.append(Http::Headers::get().VaryValues.AcceptEncoding);
```

Comment:
You can replace all the above with:
  absl::StrAppend(&headers.insertVary.value(),   ", ", Http::Headers::get()VaryValues.AcceptEncoding);
I think that will do all the right-sizing with way less code :)

Maybe you can't modify the value in place though (not sure about the Envoy header semantics off hand) in which case you may have to StrCat the whole thing rather than StrAppend the new part.